File: python/src/02_wallet.py
"""02 -- Keys, entirely offline.

Nothing here touches the network. That matters more than usual on PureChain: gas
is free, so a key created here can transact immediately with no funding step, no
faucet, and no balance.
"""

from __future__ import annotations

import asyncio
import json

from purechain import address, units, wallet
from shared import heading, note, row


async def main() -> None:
    heading("02", "Wallets, addresses and units -- no network involved")

    fresh = wallet.create()
    row("wallet.create()", fresh.address)

    generated = wallet.create_with_mnemonic()
    words = generated.mnemonic.split()
    row("mnemonic", " ".join(words[:3]) + f" … ({len(words)} words)")
    row("derived", generated.signer.address)

    # Derivation is standard BIP-44 coin type 60, so an imported phrase gives the
    # same addresses MetaMask would show.
    for i, acct in enumerate(wallet.derive_accounts(generated.mnemonic, 3)):
        row(f"  m/44'/60'/0'/0/{i}", acct.address)

    note(f"Derivation path is {wallet.DEFAULT_HD_PATH}/<index> -- the Ethereum standard.")

    # Keystore round-trip. Deliberately slow: scrypt is what makes guessing the
    # password expensive.
    print()
    keystore = wallet.to_keystore(generated.signer, "correct horse battery staple")
    row("keystore", f"v{json.loads(keystore)['version']} JSON, {len(keystore)} chars")
    restored = wallet.from_keystore(keystore, "correct horse battery staple")
    ok = "✓ matches" if restored.address == generated.signer.address else "✗"
    row("restored", f"{restored.address}  {ok}")

    # Signing and verification, still offline.
    print()
    signature = await generated.signer.sign_message("gm")
    row("sign_message", signature[:26] + "…")
    row("recovered", wallet.recover_message_address("gm", signature))
    row("verifies", wallet.verify_message("gm", signature, generated.signer.address))
    row("tampered msg", wallet.verify_message("gn", signature, generated.signer.address))

    print()
    checksummed = address.checksum(generated.signer.address)
    row("checksummed", checksummed)
    row("normalized", address.normalize(checksummed))
    row("equals (mixed)", address.equals(generated.signer.address, checksummed))

    # Units. Amounts are int everywhere -- never floats.
    print()
    row("parse_pcn('1.5')", units.parse_pcn("1.5"))
    row("format_pcn", units.format_pcn_with_symbol(units.parse_pcn("1.5")))
    total = units.parse_pcn("0.1") + units.parse_pcn("0.2")
    row("0.1 + 0.2", f"{units.format_pcn(total)}  (exact -- int, not float)")

    note("0.1 + 0.2 != 0.3 in binary floating point. On a balance that is a real loss.")


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/src/02-wallet.ts
/**
 * 02 — Keys, entirely offline.
 *
 * Nothing here touches the network. That matters more than usual on PureChain:
 * gas is free, so a key created here can transact immediately with no funding
 * step, no faucet, and no balance.
 */

import { address, units, wallet } from "purechain-sdk";
import { heading, note, row } from "./shared.ts";

export async function main(): Promise<void> {
  heading("02", "Wallets, addresses and units — no network involved");

  // A throwaway key.
  const fresh = wallet.create();
  row("wallet.create()", fresh.address);

  // A recovery phrase, and the account it derives.
  const { signer, mnemonic } = wallet.createWithMnemonic();
  row("mnemonic", mnemonic.split(" ").slice(0, 3).join(" ") + " … (12 words)");
  row("derived", signer.address);

  // Derivation is standard BIP-44 coin type 60, so an imported phrase gives the
  // same addresses MetaMask would show.
  const accounts = wallet.deriveAccounts(mnemonic, 3);
  accounts.forEach((a, i) => row(`  m/44'/60'/0'/0/${i}`, a.address));

  note(`Derivation path is ${wallet.DEFAULT_HD_PATH}/<index> — the Ethereum standard.`);

  // Keystore round-trip. Deliberately slow: scrypt is what makes guessing the
  // password expensive.
  console.log();
  const keystore = await wallet.toKeystore(signer, "correct horse battery staple");
  row("keystore", `v${JSON.parse(keystore).version} JSON, ${keystore.length} chars`);
  const restored = await wallet.fromKeystore(keystore, "correct horse battery staple");
  row("restored", `${restored.address}  ${restored.address === signer.address ? "✓ matches" : "✗"}`);

  // Signing and verification, still offline.
  console.log();
  const signature = await signer.signMessage("gm");
  row("signMessage", signature.slice(0, 26) + "…");
  row("recovered", wallet.recoverMessageAddress("gm", signature));
  row("verifies", wallet.verifyMessage("gm", signature, signer.address));
  row("tampered msg", wallet.verifyMessage("gn", signature, signer.address));

  // Address helpers.
  console.log();
  row("checksummed", address.checksum(signer.address));
  row("normalized", address.normalize(address.checksum(signer.address)));
  row("equals (mixed)", address.equals(signer.address, address.checksum(signer.address)));

  // Units. Amounts are bigint everywhere — never floats.
  console.log();
  row("parsePCN('1.5')", units.parsePCN("1.5"));
  row("formatPCN", units.formatPCNWithSymbol(units.parsePCN("1.5")));
  const sum = units.parsePCN("0.1") + units.parsePCN("0.2");
  row("0.1 + 0.2", `${units.formatPCN(sum)}  (exact — bigint, not float)`);

  note("0.1 + 0.2 !== 0.3 in binary floating point. On a balance that is a real loss.");
}

if (import.meta.url === `file://${process.argv[1]}`) await main();
